### PR TITLE
KAFKA-18049 Upgrade the caffeine version to 3.1.1

### PR DESCRIPTION
**Pull request by m1a2st**
Jira: https://issues.apache.org/jira/browse/KAFKA-18049
as title 

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by divijvaidya** [comment]
Can you also please reason about any breaking changes between 2.x and 3.x that we should know about? For your reference, there is only one cache in Kafka code base that uses Caffeine, so it should be easy to reason about it.

**Comment by mimaison** [comment]
With version 3, do we still need to exclude checker-qual? https://github.com/mimaison/kafka/blob/trunk/build.gradle#L2278-L2280

**Comment by ben-manes** [comment]
The [exclusion](https://github.com/apache/kafka/pull/16578) was a preference as not used at runtime. It is still a dependency and that pr contains details.

From your usage, I don’t recall any changes you might need to be aware of.

**Comment by m1a2st** [comment]
Kafka only use in `RemoteIndexCache`, In caffeine 3.X version, I think the change we should know is  
- CacheWriter usages can be replaced by Map computations and Caffeine.evictionListener in 3.0.0
- Update the least support version to Java11 in 3.0.0
- Fixed Map.equals when the traversal triggers an eviction and the subset of live entries matches in 3.1.0
- Improved the frequency sketch by better utilizing the cpu cache line to reduce memory accesses 3.1.2
- Improved builder construction time in 3.1.7

If I have missed anything, feel free to add or supplement further.

**Comment by m1a2st** [comment]
I will review the failed tests to determine if any changes in Caffeine 3.X version behavior caused the failures.
- FAILED ❌ RemoteIndexCacheTest > testReloadCacheAfterClose()
- FAILED ❌ RemoteIndexCacheTest > testCorrectnessForCacheAndIndexFilesWhenResizeCache()

**Comment by mimaison** [comment]
Can you explain why we're picking 3.1.1 instead of the latest version?

**Comment by m1a2st** [comment]
> Can you explain why we're picking 3.1.1 instead of the latest version?

The main difference between 3.1.1 and >=3.1.2 is the improved eviction mechanism. 
- 3.1.2:
  - Approach: Processes a single candidate entry at a time, moving sequentially in LRU order.
- 3.1.1:
  - Approach: Supports batch processing of multiple candidate entries, with movement direction controlled by the remaining candidates count.
 
It appears that this change will affect RemoteIndexCache original behavior.

**Comment by mimaison** [comment]
It would be good to add that as a comment next to the version, so we know why this specific version was picked. Because I would not expect a patch version to change behavior.

**Comment by mimaison** [comment]
@ben-manes Is the change of behavior described in https://github.com/apache/kafka/pull/17879#issuecomment-2498323317 expected?
Is it common to have breaking changes in patch releases? According to @m1a2st using caffeine > 3.1.1 breaks some of our tests. Are we not using it properly?

**Comment by ben-manes** [comment]
The eviction selection is not part of the external contract, which tries to convey that the cache is allowed a lot of leniency on its choices. The technical details are shared in engineering docs because its fun and interesting. The policy's goal is merely to maximize the hit rate which is a very fuzzy topic as every workload has different characteristics, so we try to have a balanced way to be near optimal in all cases, whereas alternative policies tend to have spectacular failure cases.

Any test that assumes a deterministic eviction order is already broken because the cache includes some randomization to [defeat](https://github.com/ben-manes/caffeine/blob/a31defdc6751136fe47cf81fdc18a70923488da5/caffeine/src/main/java/com/github/benmanes/caffeine/cache/BoundedLocalCache.java#L165-L173) denial of service attacks. For the cache's tests we work around this by [unsafely forcing](https://github.com/ben-manes/caffeine/blob/a31defdc6751136fe47cf81fdc18a70923488da5/caffeine/src/test/java/com/github/benmanes/caffeine/cache/Reset.java#L40-L44) the `ThreadLocalRandom` seeds, meaning going deeply into implementation details.

A small tuning change like that one or refactoring can have a subtle impact that does not change the net hit rate. For example if one modifies the hashing function then different counter collisions might impact frequency estimates. This could cause an inconsequential difference in the eviction choice for the overall hit rate (e.g. when both the candidate and victim entries are reasonable to evict). Please do not depend on those types of implementation details.

**Comment by showuon** [comment]
Thanks for the explanation @ben-manes !
So the improvement in version >= 3.1.2 is to avoid hash DOS attack. In that case, I think we should try to modify our tests to make them pass under version >= 3.1.2. @m1a2st , do you agree?

**Comment by ben-manes** [comment]
> So the improvement in version >= 3.1.2 is to avoid hash DOS attack.

Just to clarify that was an example for why the test could be unstable with the existing version. The protection has been in place since v2.2.7.

The commits, https://github.com/ben-manes/caffeine/commit/25b7d17b1a246a63e4991d4902a2ecf24e86d234 and https://github.com/ben-manes/caffeine/commit/46cd5833aea8190bd9cc656286f58efbde91a792, show that the intent was to better handle when an entry exceeds the capacity of the cache or a region to avoid unnecessary evictions. The walking of the LRU lists changed a bit, perhaps a bug fix iirc (I think there was a previous LRU/MRU mixup), so the evaluation changed slightly.

**Comment by mimaison** [comment]
Thanks for the clarifications!
It seems there's something wrong with our tests, I opened https://issues.apache.org/jira/browse/KAFKA-18089 to track it. In the meantime, let's upgrade to 3.1.1.

File: gradle/dependencies.gradle [hunk 1]
```diff
@@ -54,7 +54,10 @@ versions += [
   apacheds: "2.0.0-M24",
   argparse4j: "0.7.0",
   bcpkix: "1.78.1",
-  caffeine: "2.9.3", // 3.x supports JDK 11 and above
+  // The main difference between 3.1.1 and >=3.1.2 is the improved eviction mechanism,
```
Comment (mimaison) [inline gradle/dependencies.gradle]: Can you update the comment to match the discussion? You can mention KAFKA-18089. Thanks

**Comment by m1a2st** [comment]
@ben-manes, Thank you for your explanation.

**Pull request merged**